src/migrations.rs

```diff
@@ -0,0 +1,288 @@
+                    .with_context(|| {
+                        format!(
+                            "Failed to execute SQL schema version: {}",
+                            SchemaVersion::PLTSchema
+                        )
+                    })?;
+                SchemaVersion::PLTSchema
+            }
+            SchemaVersion::PLTSchema => unimplemented!(
+                "No migration implemented for database schema version {}",
+                self.as_i64()
+            ),
+        };
+        let end_time = chrono::Utc::now();
+        insert_migration(database_client, &new_version.into(), start_time, end_time).await?;
+        Ok(new_version)
+    }
+}
+
+/// Set up the migrations tables if not already there.
+async fn ensure_migrations_table(database_client: &mut DatabaseClient) -> anyhow::Result<()> {
+    let statement = "
+        CREATE TABLE IF NOT EXISTS migrations (
+            version BIGINT PRIMARY KEY,
+            description TEXT NOT NULL,
+            destructive BOOL NOT NULL,
+            start_time TIMESTAMPTZ NOT NULL,
+            end_time TIMESTAMPTZ NOT NULL)";
+
+    database_client.as_ref().execute(statement, &[]).await?;
+
+    Ok(())
+}
+
+/// Check the existence of the 'migration' table.
+async fn has_migration_table(database_client: &DatabaseClient) -> anyhow::Result<bool> {
+    let statement = "
+      SELECT EXISTS (
+            SELECT FROM information_schema.tables
+            WHERE  table_schema = 'public'
+            AND    table_name   = 'migrations'
+        );";
+
+    let row_opt = database_client.as_ref().query_opt(statement, &[]).await?;
+
+    let has_migration_table = match row_opt {
+        Some(row) => row.get::<_, bool>(0),
+        None => false,
+    };
+
+    Ok(has_migration_table)
+}
+
+/// Query the migrations table for the current database schema version.
+/// Results in an error if not migrations table found.
+pub async fn current_schema_version(
+    database_client: &DatabaseClient,
+) -> anyhow::Result<SchemaVersion> {
+    let statement = "SELECT MAX(version) FROM migrations";
+
+    let row_opt = database_client.as_ref().query_opt(statement, &[]).await?;
+    let version = match row_opt {
+        Some(row) => row.get::<_, Option<i64>>(0).unwrap_or(0),
+        None => 0,
+    };
+
+    SchemaVersion::from_version(version).context("Unknown database schema version")
+}
+
+/// Update the migrations table with a new migration.
+async fn insert_migration(
+    database_client: &mut DatabaseClient,
+    migration: &Migration,
+    start_time: chrono::DateTime<chrono::Utc>,
+    end_time: chrono::DateTime<chrono::Utc>,
+) -> anyhow::Result<()> {
+    let statement = "
+        INSERT INTO migrations (
+            version, 
+            description,
+            destructive, 
+            start_time,
+            end_time
+        )
+        VALUES ($1, $2, $3, $4, $5)
+        ON CONFLICT (version) DO UPDATE 
+            SET end_time = EXCLUDED.end_time";
```

Comment:
Handling conflicts here was a tradeoff in ccdscan in order to allow partial migrations, maybe we should just remove it and have it fail, we can introduce it if we ever need partial migrations here.